### ISSUE TITLE
RFC 3339: Support parsing Z-suffix datetimes into Zoned directly?

### ISSUE THREAD

**Issue by sharkdp** [BurntSushi/jiff#66]
RFC 3339, [section 2](https://www.rfc-editor.org/rfc/rfc3339#section-2) uses the `Z` (Zulu) suffix to denote *"a UTC offset of 00:00"*. In [section 5.7](https://www.rfc-editor.org/rfc/rfc3339#section-5.7) they go a bit further and identify "Z" with a time zone.

At the moment, jiff does not support parsing something like [`1985-04-12T23:20:50.52Z`](https://www.rfc-editor.org/rfc/rfc3339#section-5.8) into a `Zoned`. I presume that's because it interprets the `Z` suffix as a 0-offset only, not as a timezone?

I wonder if something like:
```rs
Zoned::from_str("1985-04-12T23:20:50.52Z") // Error "failed to find time zone in square brackets"
```
should be supported though? Otherwise, it feels a bit strange to me that even offset-only strings can be parsed into a `Zoned`, but `Z`-suffixed datetimes can not:
```rs
Zoned::from_str("1985-04-12T23:20:50.52[+00:00]") // Ok
```

**Comment by BurntSushi** [comment]
So to be clear here, parsing with just an offset fails as well:

```
    let result = "2024-03-09T15:00-05".parse::<Zoned>();
    eprintln!("{}", result.unwrap_err());
    let result = "2024-03-09T15:00Z".parse::<Zoned>();
    eprintln!("{}", result.unwrap_err());
```

Has this output:

```
[andrew@bart play]$ cargo -q r
failed to find time zone in square brackets in "2024-03-09T15:00-05", which is required for parsing a zoned instant
failed to find timezone 'Z' in time zone database
```

What does succeed is when you put an offset inside the time zone annotation. This comes from [RFC 9557](https://datatracker.ietf.org/doc/rfc9557/). For example, this is also valid:

```
1985-04-12T23:20:50.52+00[+00:00]
```

The idea here is that the thing in square brackets _should_ be an IANA tz identifier. If it isn't, then IMO, one should generally consider that to be undesirable behavior. But, it can arise simply because an IANA tz identifier isn't always available or because you need to interact with systems that don't support RFC 9557.

So the behavior here is, generally, a `Zoned` _requires_ that a valid TZ annotation be present. If you _don't_ have a TZ annotation, then you should generally be parsing into a `Timestamp`. And from there you can convert to a `Zoned` by providing a time zone. If you need to support _both_ datetimes with and without TZ annotations, then you should try parsing into a `Zoned` first, and if that fails, parse into a `Timestamp`.

Temporal is the thing where this behavior comes from:

* https://tc39.es/proposal-temporal/docs/zoneddatetime.html#from
* https://tc39.es/proposal-temporal/docs/strings.html

With respect to `Z`, RFC 9557 also updated RFC 3339 with this text:

```
   |  If the time in UTC is known, but the offset to local time is
   |  unknown, this can be represented with an offset of "Z".  (The
   |  original version of this specification provided -00:00 for this
   |  purpose, which is not allowed by [ISO8601:2000] and therefore is
   |  less interoperable; Section 3.3 of [RFC5322] describes a related
   |  convention for email, which does not have this problem).  This
   |  differs semantically from an offset of +00:00, which implies that
   |  UTC is the preferred reference point for the specified time.
```

Or more succinctly, also from RFC 9557:

```
   This specification updates Section 4.3 of [RFC3339], aligning it with
   the actual practice of interpreting the offset Z to mean the same as
   -00:00: "the time in UTC is known, but the offset to local time is
   unknown".
```

So parsing a datetime with `Z` into a `Zoned` is violating the intended meaning of `Z`. A `Zoned` is something where the _offset to local time from UTC is known_, but `Z` is very specifically conveying that the offset is not known. So allowing a datetime with a `Z` to be effortlessly parsed into a `Zoned` would be a footgun.

A selection of quotes from the Temporal docs linked above:

> A string like "2022-02-28T03:06:00Z" or "2022-02-28T03:06:00+02:00" lacks the time zone annotation (e.g. [America/Los_Angeles]), so it will throw an exception when parsed by Temporal.ZonedDateTime.from.
>
> A reasonable question is why Temporal.ZonedDateTime doesn't automatically use the offset (e.g. Z or +02:00) if there's no time zone suffix. The answer is that Temporal.ZonedDateTime is designed to ensure DST-safe arithmetic, which in turn requires knowing when DST starts or stops in a particular time zone. Here's a short example illustrating why this matters.
>
> ```
> function oneDayLaterInstant(s) {
>   return Temporal.ZonedDateTime.from(s).add({ days: 1 }).toInstant();
> }
> oneDayLaterInstant('2021-03-28T00:00+01:00[Europe/Paris]');
>   // => 2021-03-28T22:00:00Z
> oneDayLaterInstant('2021-03-28T00:00+01:00[+01:00]');
>   // => 2021-03-28T23:00:00Z (one hour later!)
> oneDayLaterInstant('2021-03-28T00:00+01:00');
>   // throws, because there's no time zone suffix
> ```
> The calls above use the same timestamp, but return a different result when called with a day that happens to contain a DST transition. As the second call above shows, if you really do want to do arithmetic or otherwise use Temporal.ZonedDateTime functionality, it can be done using an offset time zone suffix like [+01:00].
>
> But if there's no time zone suffix, then Temporal.ZonedDateTime.from will throw an exception. This exception has the effect of forcing developers to provide the time zone in a different way, e.g. by parsing the string into a Temporal.Instant and using its toZonedDateTimeISO() method to convert to Temporal.ZonedDateTime. This is a good outcome because whether the time zone is part of the string or called separately, requiring it enables developers to know that any Temporal.ZonedDateTime APIs will always return time-zone-aware results. If timezone-aware results are not needed, then use another type like Temporal.Instant.

OK I'll stop there. Happy to field questions.

**Comment by sharkdp** [comment]
Thank you for the detailed explanation. I think I was mainly irritated by the fact that `[+00:00]` worked, but `Z` didn't. I hadn't tried offsets without brackets.

> So the behavior here is, generally, a `Zoned` _requires_ that a valid TZ annotation be present. If you _don't_ have a TZ annotation, then you should generally be parsing into a `Timestamp`.

I understand that. I just thought that `Z` would also be a timezone. They call it a timezone in RFC 3339. `Timezone::UTC` is a thing. Specifying/sharing timestamps in UTC is a extremely common, so I had never interpreted a timestamp like `2024-07-31T07:00:00Z` as *"oh no, I don't know the local offset to UTC"*, but rather as: *"we live in an globalized world, so here's a time stamp in UTC for all of you out there (and now go ahead and transform it into **your** local timezone)"*.

Anyway, I now understand that was a misconception on my side (as clarified by the quotes from RFC 9557), so I'm going to close this.

**Comment by BurntSushi** [comment]
Yeah I had that same misconception about 6 months ago. :-) I didn't know up from down, actually.

The docs could use a lot of improvement here to be fair though.
